### DISCUSSION TITLE
Building 9 Planes to Simulate an "Infinite" Plane

### DISCUSSION THREAD

**Discussion by llh1201** [bonsai-rx/bonsai#1249] [Q&A]
Hi everyone,

I am aiming to build an extended presentation (ideally infinite) of a visual environment, for which I currently just load a large 2D texture. 

The issue when loading a 2d texture is its boundary, which the animal exceeds after very short behavior during closed-loop control.

To approach this issue in a more simple manner, we were thinking of at least extending the visual environment by loading and building 9 textures instead of 1 single (8 surrounding the center piece, see picture below). This would allow us to teleport* back the animal once it reaches a certain height and width within the texture matrix to avoid losing the visual stimulus over time. (When teleporting within the single loaded texture, we see a glitch during every reset, which would happen less often with a bigger visual environment.)

It would be super helpful if anybody could provide some instruction, on how to create such a matrix of 2d textures in more detail. I don’t know where to address this issue in the interface or the bonvision shader window programming in the backbone.

Thank you so much and best regards!
Lihao Liu

(* Teleporting: If the animal leaves the center plane at point A, it should be teleported to point A’ so that it swims into the center plane at point A’. Similarly, if the fish swims out of the center plane at point B, it should be teleported to point B’ so that it swims into the center plane at point B’.)
![9Planes](https://user-images.githubusercontent.com/91344820/217975481-394ec740-0b5a-4546-a959-197396f5f678.jpg)

**Comment by eabhorrocks** [comment]
This doesn’t quite answer your question but an alternative solution is not to move the subject/camera but move the environment/textures instead to simulate subject movement. 

To elaborate, you could have a fixed plane with a moving texture that wraps around the edges so as to appear continuous. Depending on the sort of texture you use this may or may not be easy to implement, but for simple repeating textures it works pretty well. 

Otherwise, you should be able to specify the x,y,z transform and rotation of each plane in their properties in order to set out a grid of multiple planes, unless I’m misunderstanding your question.

Edd

**Reply by llh1201** [reply]
Thank you so much for the quick reply, and this gives an excellent alternative approach! May I ask how you think such a texture movement in a fixed plane would be implemented, or if you have an examplary workflow that demonstrates what you have in mind? 

Thanks again for your advice and I really appreciate your idea!

Best regards,
Lihao Liu
